sw/device/lib/dif/dif_i2c.h

```diff
@@ -39,183 +41,265 @@ typedef enum dif_i2c_speed {
    * without going over what the target devices can handle.
    */
   dif_i2c_speed_t lowest_target_device_speed;
   /**
-   * Indicates the period of the clock driving this device, in nanoseconds.
+   * The period of the clock driving this device, in nanoseconds.
    *
    * This value should not be zero, since it is used as a divisor for
    * division.
    */
   uint32_t clock_period_nanos;
   /**
-   * Indicates the expected time it takes for the I2C bus signal to rise, in
-   * nanoseconds. This value is dependent on properties of the hardware's
-   * interconnect, and not under actual firmware control.
+   * The expected time it takes for the I2C bus signal to rise, in nanoseconds.
+   *
+   * This value is dependent on properties of the hardware's interconnect, and
+   * not under actual firmware control.
    */
   uint32_t sda_rise_nanos;
   /**
-   * Indicates the expected time for the bus singal to fall, similar to
-   * `sda_rise_nanos`.
+   * The expected time for the bus singal to fall, similar to `sda_rise_nanos`.
    */
   uint32_t sda_fall_nanos;
   /**
-   * Indicates the desired period of the SCL line, in nanoseconds. Normally,
-   * this should just be `1'000'000 / lowest_target_device_speed`, but the
-   * period may be larger if desired.
+   * The desired period of the SCL line, in nanoseconds.
+   *
+   * Normally, this should just be `1'000'000 / lowest_target_device_speed`,
+   * but the period may be larger if desired.
    *
    * Setting this value to zero will result in the minimum period being used.
    */
   uint32_t scl_period_nanos;
 } dif_i2c_timing_config_t;
 
 /**
- * Represents an I2C device. This struct should only be initialized using
- * `dif_i2c_init()`.
+ * Hardware instantiation parameters for I2C.
+ *
+ * This struct describes information about the underlying hardware that is
+ * not determined until the hardware design is used as part of a top-level
+ * design.
+ */
+typedef struct dif_i2c_params {
+  /**
+   * The base address for the I2C hardware registers.
+   */
+  mmio_region_t base_addr;
+} dif_i2c_params_t;
+
+/**
+ * Runtime configuration for I2C.
+ *
+ * This struct describes runtime timing parameters. Computing these values is
+ * somewhat complicated, so these fields should be initialized using the
+ * `dif_i2c_compute_timing()` function. A caller is, however, free to compute
+ * these values themselves if they prefer, so long as the I2C spec is
+ * respected.
+ *
+ * These values correspond to those in Table 10 of the I2C spec, and are given
+ * in units of input clock cycles.
+ */
+typedef struct dif_i2c_config {
+  uint16_t scl_time_high_cycles;
+  uint16_t scl_time_low_cycles;
+  uint16_t rise_cycles;
+  uint16_t fall_cycles;
+  uint16_t start_signal_setup_cycles;
+  uint16_t start_signal_hold_cycles;
+  uint16_t data_signal_setup_cycles;
+  uint16_t data_signal_hold_cycles;
+  uint16_t stop_signal_setup_cycles;
```

Comment:
Unless this should be treated as an opaque type, these fields should have comments (can be addressed later).

Reply:
I agree. The reason I left them out is because they're already documented in the I2C spec, and every comment would boil down to "see the I2C spec", which felt unhelpful.